Do not render report flag for reviews by the current user

SeanPrashad opened the pull request:
Fixes #6416: "Flag" should be removed from the profile page

This PR prevents the "flag" button from being rendered for reviews made by the current user. Essentially, if Sean leaves a review, he should not be allowed to flag his own comment.

**Before:**

![before_removing_flag](https://user-images.githubusercontent.com/13009507/46899863-14c1d200-ce67-11e8-8be0-886cf5cfa1e4.gif)

**After:**

![after_removing_flag](https://user-images.githubusercontent.com/13009507/46899867-18edef80-ce67-11e8-81da-4e32cb892e00.gif)

SeanPrashad commented:
For the failing tests, it looks like `siteUser` is not set for some of them since they are stubbed and I'm trying to figure out how to properly set it without touching too much!

SeanPrashad commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -562,11 +569,12 @@ describe(__filename, () => {
   it('lets you flag a developer reply', () => {
-    const { reply } = _setReviewReply();
-    const root = renderReply({ reply });
+    const review = createReviewAndSignInAsUnrelatedUser();
```
This test was giving me some trouble - not sure if this is the route for us to take. Feedback greatly appreciated!

kumar303 replied:
> This test was checking a developer's reply to an existing review. I think I changed the logic behind it - would appreciate feedback.

Why did you need to change the test? I put it back to how it was and it passed for me. Maybe it was related to not being signed in?

kumar303 commented:
I think the direction of this patch is good -- hiding the flag menu for all screens. I'd rather implement this feature consistently rather than just for the user profile page. I personally don't like hidden UI controls because it can be confusing to know what state you're in but everyone else argued against that so I concede 😁

kumar303 commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -481,7 +481,8 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
 
-        {flaggable && review ? (
+        {/* Do not render the "flag" button for replies made by the current user */}
+        {flaggable && review && siteUser && siteUser.id !== review.userId ? (
```
If the user is signed out, the *Flag* link should always be visible. Otherwise, if someone wants to flag a review, they might not know that they have to sign in first.

This is what it looks like (there is actually a style regression that I will fix right now! https://github.com/mozilla/addons-frontend/issues/6656)

<img width="1372" alt="screenshot 2018-10-16 11 45 43" src="https://user-images.githubusercontent.com/55398/47033150-b41de800-d139-11e8-99f6-87c15ac0f15d.png">

SeanPrashad commented:
I've updated the logic to **always** render the flag button if the user isn't logged in:

![post_feedback_from_kumar](https://user-images.githubusercontent.com/13009507/47112477-75b12780-d224-11e8-975a-484de9d73df6.gif)

kumar303 commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -481,7 +486,7 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
         ) : null}
 
-        {flaggable && review ? (
+        {flaggable && review && showFlagButton ? (
```
I liked how simple this was in your first patch. It was a lot easier to read. What about this?

```suggestion
        {flaggable && review && (!siteUser || siteUser.id !== review.userId) ? (
```

kumar303 replied:
You also need to add a test that makes sure the flag button is not hidden when the user is signed out. After writing the test, you can comment out the part that allowed non-signed in users to see the flag link; the test should fail. This helps you know that the test is working.

kumar303 commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -547,7 +554,7 @@ describe(__filename, () => {
   });
 
   it('lets you flag a review', () => {
-    const review = _setReview(fakeReview);
```
I think you can put this back now, too. The test passes for me.

kumar303 commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -423,6 +423,11 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
       (review.userId === siteUser.id ||
         (this.isReply() && this.siteUserCanManageReplies()));
 
+    /* Do not render the "flag" button for reviews made by the current user */
```
Prefix comments with `//` (double slashes) when they don't need to span multiple lines.

SeanPrashad commented:
Thanks for the great feedback @kumar303! ✌🏽I've reverted the other tests back and added a new test to [check for the flag button when the user wasn't authenticated](https://github.com/mozilla/addons-frontend/pull/6620/files#diff-472cf21dd9e14bab7babf939d647df8dR586). I think this should be good to 🚢

kumar303 commented:
> Not sure if it's just me, but I can't reply to the third comment you made for your review

It's not just you. Github is buggy. I copied/pasted the quote of what I was replying to because I had actually pressed the reply button on your last comment and I knew Github is buggy around this behavior (for some reason). I guess they broke it further.

kumar303 commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -576,6 +583,21 @@ describe(__filename, () => {
+    expect(renderControls(root).find(FlagReviewMenu)).toHaveLength(1);
+  });
+
+  it('hides the flag button if you wrote the review', () => {
```
When I undo your patch, this test should fail but it does not. Try to make it fail.

Here's how I undid your patch:

```diff
diff --git a/src/amo/components/AddonReviewCard/index.js b/src/amo/components/AddonReviewCard/index.js
index 13013647a..ff535434a 100644
--- a/src/amo/components/AddonReviewCard/index.js
+++ b/src/amo/components/AddonReviewCard/index.js
@@ -482,7 +482,7 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
         ) : null}
 
         {/* Do not render the "flag" button for reviews made by the current user */}
-        {flaggable && review && (!siteUser || siteUser.id !== review.userId) ? (
+        {flaggable && review ? (
           <FlagReviewMenu
             isDeveloperReply={this.isReply()}
             openerClass="AddonReviewCard-control"
```

SeanPrashad replied:
Ok - here's how I think we should proceed:

The current test:

* Signs the user in
* Creates a review
* Checks that the flag button isn't rendered

The ideal test:

* Signs the user in
* Creates a review
* Checks that the flag button isn't rendered
* **Signs the user out**
* **Checks that the flag button is rendered**

kumar303 commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -576,6 +583,21 @@ describe(__filename, () => {
     expect(root.find(FlagReviewMenu)).toHaveLength(0);
   });
 
+  it('renders the flag button when not signed in', () => {
```
Actually, I just realized that you don't need this test because we already have one:

https://github.com/mozilla/addons-frontend/blob/61424ad5a9a200c5abfe3dfda013c1fd916217fa/tests/unit/amo/components/TestAddonReviewCard.js#L556

However, I would suggest changing the spec from 'lets you flag a review' to 'shows FlagReviewMenu when signed out'

SeanPrashad commented:
Your tip made me realize that I forgot to check if the flag button would be rendered _after_ the user logged out! I've updated this in the test and removed the old helper function I had written.

kumar303 commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -576,6 +576,21 @@ describe(__filename, () => {
+  it('hides the flag button if you wrote the review', () => {
+    const originalReviewId = 123;
+    const review = signInAndDispatchSavedReview({
+      siteUserId: originalReviewId,
```
This variable name is confusing but I see what you are trying to do. You are letting the helper take over to set `siteUserId` and `reviewUserId` to the same value. Instead of letting the helper do this implicitly, your test should explicitly set these both to the same value so that the intention is clear.

SeanPrashad replied:
Yeah, I didn't want to introduce another helper that wrapped the `signInAndDispatchSavedReview()` helper so I just used it directly. I've updated the variable naming to `siteUserId` so it hopefully helps us recognize that this is an authenticated user.

SeanPrashad commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -576,6 +576,21 @@ describe(__filename, () => {
+
+    expect(renderControls(root).find(FlagReviewMenu)).toHaveLength(0);
+
+    store.dispatch(logOutUser());
```
✅  All done
